edge-modules/iotedge-diagnostics-dotnet/src/Program.cs

```diff
@@ -0,0 +1,97 @@
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Azure.Devices.Edge.Util.TransientFaultHandling;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Configuration;
+    using Newtonsoft.Json;
+    using ProxyLib.Proxy;
+    using ProxyLib.Proxy.Exceptions;
+
+    class Program
+    {
+        public static void Main(string[] args) => MainAsync(args).Wait();
+
+        static async Task MainAsync(string[] args)
+        {
+            var config = new ConfigurationBuilder().AddCommandLine(args).Build();
+            switch (args[0])
+            {
+                case "edge-agent":
+                    await EdgeAgent(config["management-uri"]);
+                    break;
+                case "iothub":
+                    await Iothub(config["hostname"], config["port"], config["proxy"]);
+                    break;
+                case "local-time":
+                    Console.WriteLine(DateTime.Now.ToUnixTimestamp());
+                    break;
+                default:
+                    throw new Exception("Invalid args");
+            }
+        }
+
+        static async Task EdgeAgent(string managementUri)
+        {
+            string modules;
+            if (managementUri.EndsWith(".sock"))
+            {
+                modules = GetSocket.GetSocketResponse(managementUri, "/modules/?api-version=2018-06-28");
+            }
+            else
+            {
+                using (var http = new HttpClient())
+                using (var response = await http.GetAsync(managementUri + "/modules/?api-version=2018-06-28"))
+                {
+                    response.EnsureSuccessStatusCode();
+                    modules = await response.Content.ReadAsStringAsync();
+                }
+            }
+
+            if (!modules.StartsWith("HTTP/1.1 200 OK"))
+            {
+                throw new Exception($"Got bad response: {modules}");
+            }
+        }
+
+        static async Task Iothub(string hostname, string port, string proxy)
+        {
+            if (proxy != null)
+            {
+                // Uses https://github.com/grinay/ProxyLib
+                var tempProxy = proxy.Split(':');
+                string proxyAddress = tempProxy[0];
+                int proxyPort = int.Parse(tempProxy[1]);
+                ProxyClientFactory factory = new ProxyClientFactory();
+                IProxyClient proxyClient = factory.CreateProxyClient(ProxyType.Http, proxyAddress, proxyPort);
+
+                // Setup timeouts
+                proxyClient.ReceiveTimeout = (int)TimeSpan.FromSeconds(60).TotalMilliseconds;
+                proxyClient.SendTimeout = (int)TimeSpan.FromSeconds(60).TotalMilliseconds;
+
+                // Get TcpClient to futher work
+                var client = proxyClient.CreateConnection(hostname, int.Parse(port));
+                client.GetStream();
```

Comment:
Need to do a TLS handshake too.

Reply:
client.GetStream(); sends a Connect request

Reply:
It creates a TCP connection to the IoT Hub through the proxy. I'm talking about the TLS handshake with the IoT Hub over that TCP connection.

Reply:
all the old module did was call `TcpStream::connect_timeout`?

Reply:
Yes, because it didn't have openssl. Now that you have a TLS stack, you might as well use it.

Reply:
I think it already might? I can write to the stream, so presumably its done the handshake too

Reply:
All you've given the `proxyClient` is the hostname and port you want it to proxy to. It doesn't know anything about what protocol you're going to use on top of the connection, so it can't possibly know that you want to run TLS over this connection, so it can't possibly have done a TLS handshake for you.

Reply:
As discussed, this may not be possible to do in C#, and is likely to be redundant with the host's TLS check anyway, so it's fine to not do it.

Reply:
Please add comments for this detail here.